GH-46193: [Flight][Format] Extend Flight Location URI Semantics (#46194)
apache/arrow: docs/source/format/Flight.rst

representation can be parsed by both implementations, as well as Go's
``net/url`` and Python's ``urllib.parse``.

.. _flight-extended-uris:

Extended Location URIs
----------------------

In addition to alternative transports, a server may also return
URIs that reference an external service or object storage location.
This can be useful in cases where intermediate data is cached as
Apache Parquet files on cloud storage or is otherwise accessible
via an HTTP service. In these scenarios, it is more efficient to be
able to provide a URI where the client may simply download the data
directly, rather than requiring a Flight service to read it back into
memory and serve it from a ``DoGet`` request.

> To avoid the complexities of Flight Clients having to implement support

Reviewer: Trivial edit (does not affect vote in process)
```suggestion
To avoid the complexities of Flight clients having to implement support
```

> for multiple different cloud storage vendors (e.g. AWS S3, Google Cloud),

Reviewer 2: This is slightly idealistic.  Every vendor could have a different signing process and, since only HTTP URIs are shared, you are putting the burden on the client to figure out how to sign the requests.

Author: the purpose here is that there's *no* burden on the client, it just makes a GET request (or uses pre-negotiated auth)

> we extend the URIs to only allow an HTTP/HTTPS uri where the client can

Reviewer: Trivial edit (does not affect vote in process)
```suggestion
we extend the URIs to only allow an HTTP/HTTPS URI where the client can
```

perform a simple GET request to download the data. Authentication can be
handled either by negotiating externally to the Flight protocol or by the

> Server sending a presigned URL that the client can make a GET request to.

Reviewer: ```suggestion
server sending a presigned URL that the client can make a GET request to.
```

> Server sending a presigned URL that the client can make a GET request to.

Reviewer 2: This section was slightly confusing to me.  You're basically saying that "Flight isn't involved in authentication" correct?

Author: Yup. Authentication for the URI should be handled outside of the scope of the flight protocol itself. Either via separate negotiation or via using presigned URIs

This should be supported by all current major cloud storage vendors, meaning
only the server needs to know the semantics of the underlying object store APIs.

When using an extended location URI, the client should ignore any
value in the ``Ticket`` field of the ``FlightEndpoint``. The
``Ticket`` is only used for identifying data in the context of a
Flight service, and is not needed when the client is directly
downloading data from an external service.

Clients should assume that, unless otherwise specified, the data is

> being returned using the :ref:`ipc-streaming-format` just as it would

Reviewer 3: Is this supposed to be a link? Is not working as seen on the latest preview docs build:
http://crossbow.voltrondata.com/pr_docs/46194/format/Flight.html#flight-extended-uris
`ipc-streaming-format` ref does not exist, should this be:
```suggestion
being returned using the :ref:`format-ipc` just as it would
```

representation can be parsed by both implementations, as well as Go's
``net/url`` and Python's ``urllib.parse``.

Extended Location URIs
----------------------

In addition to alternative transports, a server may also return
URIs that reference an external service or object storage location.
This can be useful in cases where intermediate data is cached as
Apache Parquet files on S3 or is accessible via an HTTP service. In
these scenarios, it is more efficient to be able to provide a URI
where the client may simply download the data directly, rather than
requiring a Flight service to read it back into memory and serve it
from a ``DoGet`` request. Servers should use the following URI
schemes for this situation:

+--------------------+------------------------+
| Location           | URI Scheme             |
+====================+========================+

> | Object storage (1) | s3:, gcs:, abfs:, etc. |

Reviewer 4: IMO, let's fully list out the schemes instead of relying on `etc`

For instance: is GCS gcs, or gs? (Google itself uses gs). And Azure seems to have both abfs and abfss.

Reviewer 5: The problem with listing these explicitly is that we would ideally document the full allowed URI syntax (such as allowed parameters for each supported scheme). Do we want to do that?

Reviewer 5: > Since the change is essentially about delegating data access to a different protocol, feels like it should be fine to delegate the particulars of URI syntax and such to wherever that protocol is defined as well.

That's fine for those protocols that have a [well-known URI standard](https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml). But many protocols such as S3 don't seem to, and ad-hoc S3 URI syntaxes may differ on various aspects.

For example, the Arrow C++ routine for parsing S3 URIs allows for a number of query parameters such as `region`, `endpoint_override`, `tls_verify_certificates`... Other S3 URI implementations may define other optional parameters.
https://github.com/apache/arrow/blob/d2ddee62329eb711572b4d71d6380673d7f7edd1/cpp/src/arrow/filesystem/s3fs.cc#L349-L428

Author: I would prefer to entirely avoid having to expressly specify those details in this doc. In theory it should be a "valid uri" for whichever scheme is being used. I'll update this to exhaustively specify allowed schemes.

Author: I've updated the schemes that are specified, let me know what you all think and if i should change/modify anything in there.

Reviewer 6: It seems like handling the different URI standards and implementations for cloud providers forces a tradeoff between:
- Heavy burden for implementations to support all protocols, so that clients can remain universal, or...
- No expectation for any particular scheme to be supported, but then interoperability suffers

I wonder if we're approaching this the best way. We're introducing variability for **protocol** (e.g. `s3`, `gs`) and **format** (e.g. `arrow`, `parquet`), but I don't think that `s3` and `gs` are actually different protocols. These object stores ultimately expose REST APIs, so I think the only **protocols** we're really adding here are HTTP/HTTPS.

I think that if we subtly adjust the proposal we can increase the simplicity and interoperability for clients substantially. Specifically I would propose that servers respond with `http://` or `https://` URIs, from which the client can retrieve results simply with a GET request. Auth is handled via presigned urls or mediated separately. We would continue to use `Accept` and `Content-Type` to mediate content negotiation for the data format, which works well with simple HTTP.

In this case only the server would need to know the semantics of the object store APIs it supports. For example, instead of responding with `s3://amzn-s3-demo-bucket/test2.arrow` the server would be responsible for constructing the full URL `https://amzn-s3-demo-bucket.s3.us-west-2.amazonaws.com/...`.

IMO this keeps us closer to "generic clients" but still provides server implementations with great flexibility for content delivery. It actually gives servers more control to set specific URI parameters, and prevents clients (or Arrow maintainers) from having to handle the complexity of external cloud vendors.

Reviewer 7: fwiw, seems like delta sharing (which effectively does something very similar) already works like that. According to [spec](https://github.com/delta-io/delta-sharing/blob/main/PROTOCOL.md) urls for add/remove files are exclusively https, although spec claims to support S3, ADLS, or GCS.

Reviewer: I'm also in favor of simplifying it in the way that @joellubi suggests.

Later, if the need arises, we can consider proposals of other supported URI schemes.

Reviewer 4: Seems reasonable to me.

+--------------------+------------------------+
| HTTP service   (2) | http:, https:          |
+--------------------+------------------------+

Notes:

* \(1) Any auth required should be either negotiated externally to
   Flight or should use a presigned URI.
* \(2) The client should make a GET request to the provided URI
   to retrieve the data.

When using an extended location URI, the client should ignore any
value in the ``Ticket`` field of the ``FlightEndpoint``. The
``Ticket`` is only used for identifying data in the context of a
Flight service, and is not needed when the client is directly
downloading data from an external service.

Clients should assume that, unless otherwise specified, the data is

> being returned as an Arrow IPC Stream just as it would via a ``DoGet``

Reviewer 8: How about linking to https://arrow.apache.org/docs/format/Columnar.html#ipc-streaming-format:

```suggestion
being returned as an :ref:`ipc-streaming-format` just as it would via a ``DoGet``
```

with

```diff 
diff --git a/docs/source/format/Columnar.rst b/docs/source/format/Columnar.rst
index e1603e8d8e..d60d682d76 100644
--- a/docs/source/format/Columnar.rst
+++ b/docs/source/format/Columnar.rst
@@ -1462,6 +1462,8 @@ endianness that does not match the underlying system. The reference
 implementation is focused on Little Endian and provides tests for
 it. Eventually we may provide automatic conversion via byte swapping.

+.. _ipc-streaming-format:
+
 IPC Streaming Format
 --------------------
 
@@ -1503,6 +1505,8 @@ metadata length (``0x00000000``) or closing the stream interface. We
 recommend the ".arrows" file extension for the streaming format although
 in many cases these streams will not ever be stored as files.
 
+.. _ipc-file-format:
+
 IPC File Format
 ---------------
 
```

Reviewer 3: This only for `http` service type requests, we should specify. I don't think object storage type connections expected to return IPC stream, right?
